File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem15_MaximumNumberOfVowelsInASubstringOfGivenLength.java
package org.phoenix.leetcode.challenges;

import java.util.Set;

public class Problem15_MaximumNumberOfVowelsInASubstringOfGivenLength {
    public int maxVowels(String s, int k) {
        Set<Character> vowels = Set.of('a', 'e', 'i', 'o', 'u');
        int max;
        int count = 0;
        for (int i = 0; i < k; i++) {
            if (vowels.contains(s.charAt(i))) {
                count++;
            }
        }
        max = count;
        for (int i = k; i < s.length(); i++) {
            if (vowels.contains(s.charAt(i))) {
                count++;
            }
            if (vowels.contains(s.charAt(i - k))) {
                count--;
            }
            max = Math.max(max, count);
        }
        return max;
    }
}

File: JavaScript/leetcode75/src/main/Problem15_MaximumNumberOfVowelsInASubstringOfGivenLength.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxVowels = function (s, k) {
  let set = new Set(["a", "i", "e", "o", "u"]);
  let maxCount = 0,
    count = 0;
  for (let i = 0; i < k; i++) {
    count += set.has(s[i]) ? 1 : 0;
  }
  maxCount = count;
  for (let i = k; i < s.length; i++) {
    count += set.has(s[i]) ? 1 : 0;
    count -= set.has(s[i - k]) ? 1 : 0;
    maxCount = Math.max(maxCount, count);
  }
  return maxCount;
};

module.exports = maxVowels;
